### PR TITLE
Raise on invalid :as param in integration tests

### PR DESCRIPTION
**Pull request by courtenay**
see Rails/rails#27059

### CODE REVIEW

**Comment by maclover7** [comment]
cc @kaspth, since you implemented this feature

**Comment by courtenay** [comment]
oh yeah I'm an idiot, re-pushing

**Comment by courtenay** [comment]
just burned an hour trying to figure out why `as: :xml` wouldn't work, grrrr, happy to chat about alternatives if you had a good reason for falling back to WWW encoding @kaspth

**Review by rafaelfranca — CHANGES_REQUESTED** [review]
Many tests are broken because of this.

File: actionpack/lib/action_dispatch/testing/request_encoder.rb [hunk 1]
```diff
@@ -40,7 +40,7 @@ def self.parser(content_type)
     end
 
     def self.encoder(name)
-      @encoders[name] || WWWFormEncoder
+      @encoders[name] || raise(ArgumentError, "Can't respond with unregistered encoder: #{name}. Register encoders with ActionDispatch::IntegrationTest.register_encoder(:#{name})")
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/testing/request_encoder.rb]: We need to handle nil names:

```ruby
if name
  @encoders[name] || raise(#...)
else
  WWWFormEncoder
end
```

Comment (kaspth) [inline actionpack/lib/action_dispatch/testing/request_encoder.rb]: For the error message, I'd go with:

```
Unknown encoder: can't encode the request as: #{name.inspect}. Define the encoder with `ActionDispatch::IntegrationTest.register_encoder`
```

File: actionpack/test/controller/integration_test.rb [hunk 2]
```diff
@@ -980,6 +980,12 @@ def test_registering_custom_encoder
     Mime::Type.unregister :wibble
   end
 
+  def test_encoding_with_unregistered_encoder
+    assert_raise ArgumentError do
+      post_to_foos as: :unregisterd_encoder
+    end
+  end
```
Comment (kaspth) [inline actionpack/test/controller/integration_test.rb]: Needs to test that `as: nil` doesn't raise.

**Comment by kaspth** [comment]
Heads up! I opted to expand the documentation instead. Thanks for bringing this up :)

**Pull request closed without merging**